### ISSUE TITLE
Deploy automation

### ISSUE THREAD

**Issue by divergentdave** [konklone/oversight.garden#52]
We should have better tooling for deploying changes to the site. I know little about either, but Chef or Puppet might be helpful for this. Here's my brainstorming on what tasks we should automate

- Fetching the latest code from Github
- Checking out a particular revision from git
- Running npm install (and bundle install?)
- (Optionally) creating a new ElasticSearch index and managing index aliases (see #26)
- Ingesting reports from disk into a staging index
- Starting and stopping Node.js processes, with the newer version binding to a different port and reading from a different index/index alias
- Changing and reloading nginx configuration to switch to a newer Node.js process
- Bringing Node.js and/or ElasticSearch back up should they crash
- Cleaning up old ES indexes when they are no longer in use
- Handling ElasticSearch upgrades by starting a newer cluster/server on a different port

When it comes to ElasticSearch-related changes, we could be clever and always create a new index when either tasks/elasticsearch.rake or tasks/inspectors.js changes. Dealing with ElasticSearch upgrades by switching to a new cluster is probably of lower priority than the rest of the list.

Thoughts? Suggestions?

**Comment by divergentdave** [comment]
Upon further reflection, it looks like a few of these steps are already taken care of in https://github.com/konklone/oversight.io/blob/master/fabfile.py

**Comment by konklone** [comment]
Yeah, some of this is done in Fabric. I've given your public key access to the `unitedstates` user on the box, so why don't you exercise the Fabric script to stop/start/deploy the app and see how much farther you think we need to go?

Bear in mind, Fabric requires Python 2.7 still. I manage this by creating a `virtualenv` just for Fabric, using `virtualenvwrapper`, which has Python 2.7 installed inside it. Then I `workon` that virtualenv and run `fab` commands while it's active.

**Comment by konklone** [comment]
Maybe a good first test of this is to also update the `config.yml` on the production server to have the `port` in place, after #54?

**Comment by divergentdave** [comment]
Fabric is working well, plus we've expanded the tasks to include reindexing now. Closing as done.
